oppia/oppia
# Fix#17880 Set up a reverse proxy to send auth requests to firebase

## Overview


1. This PR fixes #17880.
2. This PR does the following: Creates a proxy that redirects auth requests from <oppia_domain>/__/auth to <firebase_domain>/__/auth so that the oppia website doesn't need to access the internal storage of a another website ([doc on firebase](https://firebase.google.com/docs/auth/web/redirect-best-practices#proxy-requests) for reference.)
3. (For bug-fixing PRs only) The original bug occurred because: Browsers have started to restrict access to third party website's internal storage. Our authentication token is stored on the firebase domain and our website can't access it.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Steps to test the changes
 * This PR only adds a new endpoint so it should not break anything in production in its current state.
 * According to the firebase doc, we need to take care of the following steps:
   * Configure `authDomain` in [firebase config](https://github.com/oppia/oppia/blob/4def547d5425ff55f772ec2907885a7961ee330a/core/templates/services/auth.service.ts#L120C23-L120C23) to the root oppia domain (`oppia.org` or `oppiatestserver.org` etc.). Currently its `oppiaserver.firebaseapp.com` (new to old auth domain mapping [here](https://github.com/oppia/oppia/pull/18659/files#diff-d4ca9370f51aef2dcfdf35e910fe115aa4cfb681e40ec0342a3951c7e56146ddR32))
   * Add `https://<oppia-domain>/__/auth/handler` as an authorized URI (`Authroized redirect URIs` section in GCP OAuth settings) to allow authentication requests from our domain.
   * Add `https://<oppia-domain>` as an `Authorized domain` in firebase console.
   * ~Add the new authDomain to the SAML Assertion Consumer Service (ACS) URL.~ This is not appilcable for us because we don't use SAML for authentication.

## Proof that changes are correct
The video below shows that all requests to the `/__/auth/` endpoint are redirected to the firebase app. This is not a 30x redirect so from the client's point of view, all requests were made and handled by the oppia domain itself.

Note: After [adding oppia domain to firebase mapping](https://github.com/oppia/oppia/pull/18659/files#diff-d4ca9370f51aef2dcfdf35e910fe115aa4cfb681e40ec0342a3951c7e56146ddR32), `__/auth/` requests on localhost will not redirect to firebase because there is no associated firebase domain for it.

https://github.com/oppia/oppia/assets/31477127/7f610e60-faf1-47bb-acf3-86907802bc88



#### Proof of changes on desktop with slow/throttled network



#### Proof of changes on mobile phone



#### Proof of changes in Arabic language



## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.
